### DISCUSSION TITLE
CacheDataset: Random Transforms

### DISCUSSION THREAD

**Discussion by Gijz33** [Project-MONAI/MONAI#3301] [Q&A]
Hi!

I have a question regarding the implementation I use of monai.data.CachDataset (snipped of the code can be seen below):
Are the randomized transforms applied again on the data before every new epoch?
In other words: it is possible that a specific image is flipped and gausssian noise is apllied in the first epoch, and in the second epoch non of those transforms happen (original image is used)?

I also have a question about the random transformations within an epoch. Is the transformations between images within the same epoch also random? Or is exactly the same rotation/flipping/gaussian noise applied on all the images within the same epoch?

```python
train_ds = monai.data.CacheDataset(data=train_files, transform=train_transforms)
    
    
train_loader = monai.data.DataLoader(
    train_ds,
    batch_size=batch_size,
    shuffle=True,
    num_workers=2,
    pin_memory=torch.cuda.is_available(),
)

trainer = monai.engines.SupervisedTrainer(
        device=device,
        max_epochs=max_epochs,
        train_data_loader=train_loader,
        network=net,
        optimizer=opt,
        loss_function=DiceCELoss(),
        inferer=get_inferer(),
        post_transform=train_post_transforms,
        key_train_metric={"train_mean_dice": MeanDice(include_background=False, output_transform=lambda x: (x["pred"], x["label"]))}, 
        train_handlers=train_handlers,
        amp=amp,
    )
```

**Comment by ebrahimebrahim** [comment]
> it is possible that a specific image is flipped and gausssian noise is apllied in the first epoch, and in the second epoch non of those transforms happen (original image is used)?

Yes, assuming you set `prob` to something between `0` and `1` when you created those transforms. `prob` is the probability of applying the random transform at all.

> I also have a question about the random transformations within an epoch. Is the transformations between images within the same epoch also random? Or is exactly the same rotation/flipping/gaussian noise applied on all the images within the same epoch?

The former; it's randomized for each individual data item

**Comment by Nic-Ma** [comment]
Hi @Gijz33 ,

Thanks for your interest here.
About your questions:
1. MONAI CacheDataset caches the result before the first random transform of the chain, then reuse the exact same cache content as start point to apply the following random transforms for every epoch.
2. The random transform behaviors are different for every image.

You can find more details from the doc-string:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/dataset.py#L520

Thanks.

**Comment by Gijz33** [comment]
Thank you @Nic-Ma and @ebrahimebrahim for your clear answers!
